### DISCUSSION TITLE
typescript-react-apollo should also generate SDK

### DISCUSSION THREAD

**Discussion by cyberhck** [dotansimha/graphql-code-generator#6238] [Q&A]
Hey everyone,
I'm very new to this project and I'm having an issue.

I have some scenario where I need to make an API call outside of a react components (example: callbacks, async validation etc)

while callbacks can just use the lazy version, it's still not good, currently it returns the data back to original components.

I'm looking for a way to use something like graphql-request but with apollo client, because I still want the hooks, but I want additional operations support.

Is that already possible? or is my only hope to have two graphql client in a single project?

I could use the typescript-generic-sdk, apollo client being so popular I was hoping I could just set a flag which also generates SDK on typescript-react-apollo plugin.

Am I missing something? Any help would be appreciated :)

Thanks in advance.

**Comment by ardatan** [comment]
React Apollo plugin has a completely different approach comparing to generic-sdk. One generates React hooks and the other one generates vanilla-js framework independent functions/methods that return Promises or AsyncIterables.
Generated SDK by generic-sdk plugin can take any kind of "sdkRequester" so you can call any client under the hood.

**Reply by cyberhck** [reply]
Is there a pre-built version of generic-sdk or some pre-published package which implements requester using apollo client? I think I saw one gist somewhere, but I don't feel like copying and pasting that, I'd have to do it if there was no other choice, but it doesn't make sense for already generated code to actually not allow that.

Shouldn't this allow me to generate hook as well as sdk?

**Reply by cyberhck** [reply]
if I were to send a PR which generates sdk, would this project accept?

**Reply by cyberhck** [reply]
Another reason I want react-apollo to generate sdk is that it'll also support subscriptions :slightly_smiling_face:

**Reply by ardatan** [reply]
I see but that's out of scope of react-apollo package. Maybe it is better to have some documentation that shows how to use apollo-client with generic-sdk.

**Reply by cyberhck** [reply]
but making subscriptions work with generic-sdk using apollo client is going to be very tricky, isn't it?

while just generating sdk is quite easy.

**Reply by cyberhck** [reply]
also, how's that out of scope with react-apollo package? not all projects can use hooks, so they all have to use generic-sdk? how will subscriptions work?

**Reply by ardatan** [reply]
If the project doesn't use hooks, components or HOCs, it doesn't need React-Apollo at all so that project doesn't neexd ReactApollo plugin.
This kind of projects can use Apollo Client directly created by using the Apollo's core package then that client can be used with the SDK generated with generic-sdk plugin.

Also about subscriptions, generic-sdk can have AsyncIterable/Observable as a result type.
@dotansimha Am I missing something?

**Reply by cyberhck** [reply]
are you saying apollo client shouldn't be used without hooks?

Also, is that really a good developer experience? generate generic sdk, then having to write a bridge from apollo client to generic sdk? A good developer experience will need someone to implement the requester implementation.

What's the downside of adding a flag which generates sdk side by side of hooks?

All the docs and types are generated there anyway, it's just 1 object with multiple keys and it's something opt-in, so why not?

**Reply by ardatan** [reply]
> are you saying apollo client shouldn't be used without hooks?

I am saying the opposite. 
Let's make something clear. React-Apollo isn't equal to Apollo Client, right? React-Apollo is the integration of Apollo Client for React like Apollo-Angular is for Angular, Vue-Apollo for Vue etc.
And you want to have a generated SDK that uses Apollo Client not React-Apollo. So what if someone just wants to have this SDK and they don't use React at all. Isn't that weird to use a plugin called react-apollo in order to generate a framework independent SDK? However, instead of having those inside react-apollo plugin, I am OK to discuss about having another plugin something like "apollo-client-sdk".  

> All the docs and types are generated there anyway, it's just 1 object with multiple keys and it's something opt-in, so why not?

The types needed for SDK are generated by operation plugin not react-apollo plugin just like all other client plugins such as apollo-angular, generic-sdk and vue-apollo.

**Reply by cyberhck** [reply]
hmm, I didn't know that, being very new to gql and gql-code-generator ecosystems,

so from what I understood, react-apollo only generates hooks and hoc or data components,

BUT I'd be able to write a plugin which generates apollo client sdk but I can use both plugins to output to same file without having duplicate declarations?

**Reply by cyberhck** [reply]
hi @ardatan , thanks for the info, I've implemented #6242 but I actually agree with you. If you can help me a little I'd love to send another PR which does the sdk part.

My few questions would be: how to generate the variable types and graphql documents? is it enough to define a plugin for this?
which existing plugin should I look at for this?

also I wish to add a hook which does a useApolloClient and returns the SDK, is it fine to add that here? (by using a config to turn on)

**Reply by dotansimha** [reply]
Just saw this discussion here. I agree with @ardatan here. 

I'm not sure if it's a good idea to create a single hook that returns the SDK - because each change in that hook might lead to a reload of all components that use it, regardless if they use the modified operation.

**Reply by cyberhck** [reply]
I think it's fine, the people who would use:
```typescript
const client = useApolloClient();
const sdk = getSdk(client);
```
Could just write a hook themselves:

```typescript
const useSdkHook = () => {
  return getSdk(useApolloClient());
};
```

Once, then just re-use it again and again :)

**Reply by dotansimha** [reply]
> I think it's fine, the people who would use:
> 
> ```ts
> const client = useApolloClient();
> const sdk = getSdk(client);
> ```
> 
> Could just write a hook themselves:
> 
> ```ts
> const useSdkHook = () => {
>   return getSdk(useApolloClient());
> };
> ```
> 
> Once, then just re-use it again and again :)

I have to admit the using `typescript-react-apollo` generated hooks (specific one per operation) seems simpler for me. Also `typed-document-node` improves DX by a lot. I'm not sure I see a valid use-case in using SDK this way?

**Reply by cyberhck** [reply]
I'm using Yup to validate a schema, for one of the validations, I'll need to call my graphql server.

Now because this is in a separate call, I'll need non-hook way to call SDK, but I don't want to type:
```typescript
const client = useApolloClient();
client.query<ValidateSomethingQuery, ValidateSomethingVariables>({query: ValidateSomethingDocument, variables: {value: "something"});
```

what I'd like to type:
```typescript
const sdk = getSdk(useApolloClient());
const response = await sdk.validateSomething({value: "something"});
```
Second is much easier to use.

**Reply by dotansimha** [reply]
Why not just:

```
const response = useValidateSomething({value: "something"});

// useValidateSomething is generated from typescript-react-apollo
```

?

**Reply by cyberhck** [reply]
https://formik.org/docs/guides/validation

Because I use yup validation.

Yup validation let's you add a test function which can return a promise.

Which isn't inside a react component, but a separate .ts file.

**Reply by cyberhck** [reply]
I also want my business logic kind of separate from my UI logic, while hooks do have their places in Container Components, I'm not using them everywhere, I'm handling the data using react-query. I'm not completely abandoning them either, in some pages where I have no use for react-query, I'm still making those API calls using hooks.

**Reply by cyberhck** [reply]
for cases where people want to use apollo client without a framework (vanillaJS OR with react but without hooks), I've added #6246
